File: src/bumper_bot_py/bumper_bot_py/simple_turtlesim_kinematics.py
from rclpy.node import Node
import rclpy
from turtlesim.msg import Pose
from geometry_msgs.msg import Twist
import math


class SimpleTurtleSimKinematics(Node):
    def __init__(self):
        super().__init__("simple_turtle_sim_kinematics")


        # Create a subscriber for the turtle1 pose
        self.turtle1_pose_sub_ = self.create_subscription(
            Pose,"/turtle1/pose",self.turtle1_pose_callback, 10 )
        
        
        #create a subscriber for the turtle2 pose
        self.turtle2_pose_sub_ = self.create_subscription(
            Pose,"/turtle2/pose", self.turtle2_pose_callback,10 )


        self.last_turtle1_pose = Pose()
        self.last_turtle2_pose = Pose()

    
    def turtle1_pose_callback(self, msg: Pose):
        # Store the last pose of turtle1
        self.last_turtle1_pose = msg

    def turtle2_pose_callback(self, msg: Pose):
        # Store the last pose of turtle1
        self.last_turtle2_pose = msg

        Tx = self.last_turtle2_pose.x - self.last_turtle1_pose.x
        Ty = self.last_turtle2_pose.y - self.last_turtle1_pose.y

        # Calculate the angle difference
        theta = self.last_turtle1_pose.theta - self.last_turtle2_pose.theta
        theta_deg  = theta * 180 / 3.14

        self.get_logger().info(""" \n
               Translation from turtle1 to turtle2: \n 
               Tx : %f  \n 
               Ty : %f  \n
               Rotation matrix t1->t2 (rad) : %f \n
               Rotation matrix t1->t2 (deg) : %f \n 
               |R11     R12| : |%f       %f| \n
               |R21     R22| : |%f       %f| \n """ % (Tx, Ty, theta, theta_deg , math.cos(theta) , -math.sin(theta), 
                                                                                  math.sin(theta) , math.cos(theta)))

def main(args=None):
    rclpy.init()
    node = SimpleTurtleSimKinematics()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()



        

File: src/bumper_bot_cpp/src/simple_turtlesim_kinematics.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <turtlesim/msg/pose.hpp>



class SimpleTurtleSimKinematics: public rclcpp::Node
{
    public:
        SimpleTurtleSimKinematics() : Node("simple_turtlesim_kinematics")
        {

            t1_pose_subscriber_ = this->create_subscription<turtlesim::msg::Pose>(
                "/turtle1/pose", 10, std::bind(&SimpleTurtleSimKinematics::t1_pose_callback, this, std::placeholders::_1));

            t2_pose_subscriber_ = this->create_subscription<turtlesim::msg::Pose>(
                "/turtle2/pose", 10, std::bind(&SimpleTurtleSimKinematics::t2_pose_callback, this, std::placeholders::_1));
            

        }
    private:

        rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr t1_pose_subscriber_;
        rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr t2_pose_subscriber_;

        float t1_x_, t1_y_;
        float t2_x_, t2_y_;
        float t_vx_, t_vy_;
        float t1_theta_, t2_theta_;
        float rad_theta_, deg_theta_;


        void t1_pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
        {
            t1_x_ = msg->x;
            t1_y_ = msg->y;
            t1_theta_ = msg->theta;

        }
        void t2_pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
        {
           t2_x_ = msg->x;
           t2_y_ = msg->y;
           t2_theta_ = msg->theta;

              t_vx_ = t2_x_ - t1_x_;
              t_vy_ = t2_y_ - t1_y_;

              rad_theta_ = t1_theta_ - t2_theta_;
              deg_theta_ = rad_theta_ * (180.0 / 3.14159265358979323846);


                RCLCPP_INFO(this->get_logger(), "Translation from T2 to T1: (%.2f, %.2f)", t_vx_, t_vy_);
                RCLCPP_INFO_STREAM(this->get_logger(), 
                                "Rotation MArtix from T1 -T2:\n"<<
                                 "|R11       R12|:   |"<< std::cos(rad_theta_) <<"\t"<< -std::sin(rad_theta_) <<" | \n"<<
                                 "|R21       R22|:   |"<< std::sin(rad_theta_) <<"\t"<< std::cos(rad_theta_) <<" | \n");
        }


};

//main function
int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleTurtleSimKinematics>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}